### PR TITLE
Update direct access behaviour

### PR DESCRIPTION
**Pull request by mickmak-busicat**
ref dev-1822

Preview:

## without custom domain

https://github.com/user-attachments/assets/e8745ff9-0858-4c61-8d82-13db35d86cfc

## added custom domain

https://github.com/user-attachments/assets/856d1570-b06f-4562-aab1-260b17dfc2ac

### CODE REVIEW

**Comment by mickmak-busicat** [comment]
@louischan-oursky how should we handle different language for new translation string? from google translate or need to translate them in some paid service
let me know if i also need to change them🙏

File: pkg/lib/config/ui.go [hunk 1]
```diff
@@ -37,8 +39,11 @@ type UIConfig struct {
 	DarkThemeDisabled      bool              `json:"dark_theme_disabled,omitempty"`
 	LightThemeDisabled     bool              `json:"light_theme_disabled,omitempty"`
 	WatermarkDisabled      bool              `json:"watermark_disabled,omitempty"`
+	DirectAccessDisabled   bool              `json:"direct_access_disabled,omitempty"`
 	// client_uri to use when client_id is absent.
 	DefaultClientURI string `json:"default_client_uri,omitempty"`
+	// branding_page_uri to use when client_id is absent.
+	DefaultBrandingPageURI string `json:"default_branding_page_uri,omitempty"`
```
Comment (louischan-oursky) [inline pkg/lib/config/ui.go:46]:
```
// brand_page_uri is shown when the UI has direct_access_disabled.
BrandPageURI string `json:"brand_page_uri,omitempty"`
```

That is, please rename this config to `brand_page_uri` and update the comment.

File: resources/authgear/templates/de/translation.json [hunk 2]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Passwort verbergen",
   "return-page-title": "Zurück zu Ihrem vorherigen Standort",
   "return-page-description": "Bitte kehren Sie zu Ihrem vorherigen Standort zurück, um fortzufahren. Wenn Sie mit einer mobilen App interagiert haben, wechseln Sie bitte wieder zu dieser App. Wenn Sie mit einer Website interagiert haben, wechseln Sie bitte wieder zu diesem Tab.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/de/translation.json:463]: Remove this. This is supposed to be generated.

File: resources/authgear/templates/el/translation.json [hunk 3]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Απόκρυψη κωδικού πρόσβασης",
   "return-page-title": "Επιστροφή στο σημείο που βρισκόσασταν",
   "return-page-description": "Παρακαλούμε επιστρέψτε στο σημείο που βρισκόσασταν για να συνεχίσετε. Αν αλληλεπιδρούσατε με μια εφαρμογή κινητού, παρακαλούμε επιστρέψτε σε αυτήν. Αν αλληλεπιδρούσατε με έναν ιστότοπο, παρακαλούμε επιστρέψτε σε αυτήν την καρτέλα.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/el/translation.json:463]: Remove this.

File: resources/authgear/templates/en/web/authflowv2/direct_access_disable.html [hunk 4]
```diff
@@ -0,0 +1,28 @@
+{{ template "authflowv2/__direct_access_disable_page_frame.html" . }}
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/direct_access_disable.html:1]: The filename should be `direct_access_disabled.html`

File: resources/authgear/templates/en/web/authflowv2/direct_access_disable.html [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        (dict
+          "LogoClass" "brand-logo--settings"
+        )
+        .
+      )
+    }}
+  </div>
+  
+  <div class="text-center px-6 tablet:px-8 desktop:px-8 max-w-90 mx-auto">
+    <p class="text-3xl break-words body-text--lg font-bold mb-4">{{ $.Translations.RenderText "app.name" nil }}</p>
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/direct_access_disable.html:18]: Use `{{ translate "app.name" }}`

File: resources/authgear/templates/en/web/authflowv2/direct_access_disable.html [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    {{ if $.DefaultBrandingPageURI }}
+      <p class="body-text--md">
+        <a href="{{ $.DefaultBrandingPageURI }}">{{ $.DefaultBrandingPageURI }}</a>
+      </p>
+    {{ else }}
+      <p class="break-words body-text--sm m-0">{{ template "configure-direct-access-description" }}</p>
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/direct_access_disable.html:24]: Use `{{ translate "configure-direct-access-description" }}`

File: resources/authgear/templates/es-419/translation.json [hunk 7]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Ocultar contraseña",
   "return-page-title": "Volver a donde estabas",
   "return-page-description": "Por favor, vuelve a donde estabas para continuar. Si estabas interactuando con una aplicación móvil, por favor, vuelve a la aplicación. Si estabas interactuando con un sitio web, por favor, vuelve a esa pestaña.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/es-419/translation.json:463]: Remove this.

File: resources/authgear/templates/es-ES/translation.json [hunk 8]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Ocultar contraseña",
   "return-page-title": "Volver a donde estabas",
   "return-page-description": "Por favor, vuelve a donde estabas para continuar. Si estabas interactuando con una aplicación móvil, vuelve a la aplicación. Si estabas interactuando con un sitio web, vuelve a esa pestaña.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/es-ES/translation.json:463]: Remove this.

File: resources/authgear/templates/es/translation.json [hunk 9]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Ocultar contraseña",
   "return-page-title": "Volver a donde estabas",
   "return-page-description": "Por favor, vuelve a donde estabas para continuar. Si estabas interactuando con una aplicación móvil, por favor, vuelve a la aplicación. Si estabas interactuando con un sitio web, por favor, vuelve a esa pestaña.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/es/translation.json:463]: Remove this.

File: resources/authgear/templates/fil/translation.json [hunk 10]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Itago ang password",
   "return-page-title": "Bumalik sa kung saan ka naroroon",
   "return-page-description": "Mangyaring bumalik sa kung saan ka naroroon upang magpatuloy. Kung nakikipag-ugnayan ka sa isang mobile app, mangyaring lumipat pabalik sa app. Kung nakikipag-ugnayan ka sa isang website, mangyaring lumipat pabalik sa tab na iyon.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/fil/translation.json:463]: Remove this.

File: resources/authgear/templates/fr/translation.json [hunk 11]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Masquer le mot de passe",
   "return-page-title": "Retourner à l''endroit où vous étiez",
   "return-page-description": "Veuillez retourner à l''endroit où vous étiez pour continuer. Si vous interagissiez avec une application mobile, veuillez revenir à l''application. Si vous interagissiez avec un site Web, veuillez revenir à cet onglet.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/fr/translation.json:463]: Remove this.

File: resources/authgear/templates/id/translation.json [hunk 12]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Sembunyikan kata sandi",
   "return-page-title": "Kembali ke tempat Anda berada",
   "return-page-description": "Silakan kembali ke tempat Anda berada untuk melanjutkan. Jika Anda berinteraksi dengan aplikasi seluler, silakan beralih kembali ke aplikasi tersebut. Jika Anda berinteraksi dengan situs web, silakan beralih kembali ke tab tersebut.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/id/translation.json:463]: Remove this.

File: resources/authgear/templates/it/translation.json [hunk 13]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Nascondi password",
   "return-page-title": "Torna dove eri",
   "return-page-description": "Per favore, torna dove eri per continuare. Se stavi interagendo con un''app mobile, per favore torna all''app. Se stavi interagendo con un sito web, per favore torna a quella scheda.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/it/translation.json:463]: Remove this.

File: resources/authgear/templates/ja/translation.json [hunk 14]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "パスワードを隠す",
   "return-page-title": "元の場所に戻る",
   "return-page-description": "続行するには元の場所に戻ってください。モバイルアプリと連携していた場合はアプリに切り替え、ウェブサイトと連携していた場合はそのタブに切り替えてください。",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/ja/translation.json:463]: Remove this.

File: resources/authgear/templates/ko/translation.json [hunk 15]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "비밀번호 숨기기",
   "return-page-title": "이전 페이지로 돌아가기",
   "return-page-description": "계속하려면 이전 페이지로 돌아가세요. 모바일 앱과 상호 작용하고 있었다면 앱으로 돌아가세요. 웹사이트와 상호 작용하고 있었다면 해당 탭으로 돌아가세요.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/ko/translation.json:463]: Remove this.

File: resources/authgear/templates/ms/translation.json [hunk 16]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Sembunyikan kata laluan",
   "return-page-title": "Kembali ke tempat anda berada",
   "return-page-description": "Sila kembali ke tempat anda berada untuk meneruskan. Jika anda berinteraksi dengan aplikasi mudah alih, sila beralih kembali ke aplikasi tersebut. Jika anda berinteraksi dengan laman web, sila beralih kembali ke tab tersebut.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/ms/translation.json:463]: Remove this.

File: resources/authgear/templates/nl/translation.json [hunk 17]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Wachtwoord verbergen",
   "return-page-title": "Keer terug naar waar u was",
   "return-page-description": "Ga terug naar waar u was om door te gaan. Als u met een mobiele app bezig was, schakel dan terug naar de app. Als u met een website bezig was, schakel dan terug naar die tab.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/nl/translation.json:463]: Remove this.

File: resources/authgear/templates/pl/translation.json [hunk 18]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Ukryj hasło",
   "return-page-title": "Wróć tam, gdzie byłeś",
   "return-page-description": "Wróć tam, gdzie byłeś, aby kontynuować. Jeśli korzystałeś z aplikacji mobilnej, przełącz się z powrotem na aplikację. Jeśli korzystałeś ze strony internetowej, przełącz się z powrotem na tę kartę.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/pl/translation.json:463]: Remove this.

File: resources/authgear/templates/pt-BR/translation.json [hunk 19]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Ocultar senha",
   "return-page-title": "Voltar para onde você estava",
   "return-page-description": "Por favor, volte para onde você estava para continuar. Se você estava interagindo com um aplicativo móvel, por favor, volte para o aplicativo. Se você estava interagindo com um site, por favor, volte para essa guia.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/pt-BR/translation.json:463]: Remove this.

File: resources/authgear/templates/pt-PT/translation.json [hunk 20]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Ocultar senha",
   "return-page-title": "Voltar para onde você estava",
   "return-page-description": "Por favor, volte para onde você estava para continuar. Se você estava interagindo com um aplicativo móvel, por favor, volte para o aplicativo. Se você estava interagindo com um site, por favor, volte para essa guia.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/pt-PT/translation.json:463]: Remove this.

File: resources/authgear/templates/pt/translation.json [hunk 21]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Ocultar senha",
   "return-page-title": "Retornar de onde você estava",
   "return-page-description": "Por favor, retorne de onde você estava para continuar. Se você estava interagindo com um aplicativo móvel, por favor, volte para o aplicativo. Se você estava interagindo com um site, por favor, volte para aquela guia.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/pt/translation.json:463]: Remove this.

File: resources/authgear/templates/th/translation.json [hunk 22]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "ซ่อนรหัสผ่าน",
   "return-page-title": "กลับไปที่ที่คุณอยู่",
   "return-page-description": "โปรดกลับไปที่ที่คุณอยู่เพื่อดำเนินการต่อ หากคุณกำลังโต้ตอบกับแอปมือถือ โปรดสลับกลับไปที่แอป หากคุณกำลังโต้ตอบกับเว็บไซต์ โปรดสลับกลับไปที่แท็บนั้น",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/th/translation.json:463]: Remove this.

File: resources/authgear/templates/vi/translation.json [hunk 23]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "Ẩn mật khẩu",
   "return-page-title": "Quay lại nơi bạn đang ở",
   "return-page-description": "Vui lòng quay lại nơi bạn đang ở để tiếp tục. Nếu bạn đang tương tác với một ứng dụng di động, vui lòng chuyển lại sang ứng dụng. Nếu bạn đang tương tác với một trang web, vui lòng chuyển lại sang tab đó.",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/vi/translation.json:463]: Remove this.

File: resources/authgear/templates/zh-CN/translation.json [hunk 24]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "隐藏密码",
   "return-page-title": "返回到您之前的位置",
   "return-page-description": "请返回到您之前的位置以继续。如果您正在与移动应用程序交互,请切换回该应用程序。如果您正在与网站交互,请切换回该选项卡。",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/zh-CN/translation.json:463]: Remove this.

File: resources/authgear/templates/zh-HK/translation.json [hunk 25]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "隱藏密碼",
   "return-page-title": "返回你原來的位置",
   "return-page-description": "請返回以繼續。如你正使用移動應用程式，請返回該應用程式；如你正使用網頁，請返回該分頁。",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/zh-HK/translation.json:463]: You need to provide a translation for this. Maybe ask Fung for the copy.

File: resources/authgear/templates/zh-TW/translation.json [hunk 26]
```diff
@@ -460,6 +460,7 @@
   "hide-password-label": "隱藏密碼",
   "return-page-title": "返回你原來的位置",
   "return-page-description": "請返回以繼續。如你正使用移動應用程式，請返回該應用程式；如你正使用網頁，請返回該分頁。",
+  "configure-direct-access-description": "Direct access of the endpoint URL is not allowed. There could be a misconfiguration in the system. Please contact the systems admin.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/zh-TW/translation.json:463]: Copy the zh-HK translation here (when it is ready)

File: pkg/auth/handler/webapp/auth_entry_point_middleware.go [hunk 27]
```diff
@@ -40,12 +40,12 @@ func (m *AuthEntryPointMiddleware) Handle(next http.Handler) http.Handler {
 		host := httputil.GetHost(r, bool(m.TrustProxy))
 		isDefaultDomain := m.AppHostSuffixes.CheckIsDefaultDomain(host)
 
-		if userID != nil && !fromAuthzEndpoint {
+		if userID != nil && !fromAuthzEndpoint && !m.UIConfig.DirectAccessDisabled {
```
Comment (louischan-oursky) [inline pkg/auth/handler/webapp/auth_entry_point_middleware.go:43]: The conditions here should be

```
if userID != nil && !fromAuthzEndpoint {
  // Original block 1
} else if user == nil && !fromAuthzEndpoint && (isDefaultDomain || (!isDefaultDomain && m.UIConfig.DirectAccessDisabled)) {
  // Original block 2
} else {
  // Original block 3
}
```

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by pkong-ds** [comment]
@mickmak-busicat dailo! maybe this can help https://github.com/authgear/authgear-server/blob/main/CONTRIBUTING.md#generate-translation; just need to ask fung for `Anthropic (Claude) key`

File: pkg/auth/handler/webapp/auth_entry_point_middleware.go (hunk 27 above)
Comment (mickmak-busicat) [reply pkg/auth/handler/webapp/auth_entry_point_middleware.go:43]: fixed the condition

File: resources/authgear/templates/en/web/authflowv2/direct_access_disable.html (hunk 5 above)
Comment (mickmak-busicat) [reply resources/authgear/templates/en/web/authflowv2/direct_access_disable.html:18]: updated to use translate

File: resources/authgear/templates/en/web/authflowv2/direct_access_disable.html (hunk 6 above)
Comment (mickmak-busicat) [reply resources/authgear/templates/en/web/authflowv2/direct_access_disable.html:24]: updated to use `translate`

File: resources/authgear/templates/en/web/authflowv2/direct_access_disable.html (hunk 4 above)
Comment (mickmak-busicat) [reply resources/authgear/templates/en/web/authflowv2/direct_access_disable.html:1]: renamed

File: pkg/lib/config/ui.go (hunk 1 above)
Comment (mickmak-busicat) [reply pkg/lib/config/ui.go:46]: renamed all related config name

**Comment by louischan-oursky** [comment]
I see the comments were resolved but the PR was not updated.

File: resources/authgear/templates/zh-TW/translation.json (hunk 26 above)
Comment (mickmak-busicat) [reply resources/authgear/templates/zh-TW/translation.json]: updated to new wording

File: resources/authgear/templates/zh-HK/translation.json (hunk 25 above)
Comment (mickmak-busicat) [reply resources/authgear/templates/zh-HK/translation.json]: updated to new wording

**Comment by mickmak-busicat** [comment]
@louischan-oursky pushed update :adore2x:

File: pkg/lib/config/ui.go [hunk 28]
```diff
@@ -42,8 +42,8 @@ type UIConfig struct {
 	DirectAccessDisabled   bool              `json:"direct_access_disabled,omitempty"`
 	// client_uri to use when client_id is absent.
 	DefaultClientURI string `json:"default_client_uri,omitempty"`
-	// branding_page_uri to use when client_id is absent.
-	DefaultBrandingPageURI string `json:"default_branding_page_uri,omitempty"`
+	// brand_page_uri is shown when the UI has direct_access_disabled.
+	BrandingPageURI string `json:"branding_page_uri,omitempty"`
```
[the remark this answers is not in the document]
Comment (mickmak-busicat) [reply pkg/lib/config/ui.go:46]: sorry big eyes
small change fixed up 🙏

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
SCR-20240912-sdip
 Cannot save even I edit the post login redirect uri.

SCR-20240912-sdjx
When I select brand page, it asks me to fill in the disabled post login redirect uri field.

**Comment by mickmak-busicat** [comment]
@louischan-oursky updated changing the option will restore its initial state value for the input

https://github.com/user-attachments/assets/078a7091-cdd8-40a9-8bb2-b449026d42e3


https://github.com/user-attachments/assets/a2a41dea-054b-4f69-9f65-cdd81dc0abd1

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
1. The form depends on the currently activated domain. If the currently activated domain is the default domain, then the Custom option is disabled.

**Comment by mickmak-busicat** [comment]
@louischan-oursky updated to check current activated domain is custom for the direct access option


https://github.com/user-attachments/assets/dc41e834-d214-4294-ae4a-cc764ac1cb15

File: pkg/lib/config/testdata/config_tests.yaml [hunk 29]
```diff
@@ -605,6 +607,7 @@ config:
     public_origin: http://test
   ui:
     default_client_uri: nonsense
+    branding_page_uri: nonsense
```
Comment (louischan-oursky) [inline pkg/lib/config/testdata/config_tests.yaml:610]: @mickmak-busicat The commit for fixing the name is missing!

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/lib/config/testdata/config_tests.yaml (hunk 29 above)
Comment (mickmak-busicat) [reply pkg/lib/config/testdata/config_tests.yaml]: @louischan-oursky added back :adore2x:

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**